pkg/server/gateway.go

```diff
@@ -0,0 +1,151 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/aserto-dev/certs"
+	"github.com/aserto-dev/go-directory/aserto/directory/exporter/v2"
+	"github.com/aserto-dev/go-directory/aserto/directory/importer/v2"
+	"github.com/aserto-dev/go-directory/aserto/directory/reader/v2"
+	"github.com/aserto-dev/go-directory/aserto/directory/writer/v2"
+	metrics "github.com/aserto-dev/go-http-metrics/middleware/grpc"
+	"github.com/aserto-dev/logger"
+	"github.com/aserto-dev/topaz/pkg/builder"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"github.com/pkg/errors"
+	"github.com/rs/cors"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+var (
+	allowedOrigins = []string{
+		"http://localhost",
+		"http://localhost:*",
+		"https://localhost",
+		"https://localhost:*",
+		"http://127.0.0.1",
+		"http://127.0.0.1:*",
+		"https://127.0.0.1",
+		"https://127.0.0.1:*",
+	}
+)
+
+func NewGatewayServer(services []string, serviceConfig *builder.API, log *zerolog.Logger) (*http.Server, error) {
+	newLogger := log.With().Str("source", "http-gateway").Logger()
+
+	c := cors.New(cors.Options{
+		AllowedHeaders: []string{"Authorization", "Content-Type", "Depth"},
+		AllowedOrigins: append(allowedOrigins, serviceConfig.Gateway.AllowedOrigins...),
+		AllowedMethods: []string{http.MethodGet, http.MethodPost, http.MethodHead, http.MethodDelete, http.MethodPut,
+			http.MethodPatch, "PROPFIND", "MKCOL", "COPY", "MOVE"},
+		Debug: true,
+	})
+	c.Log = &newLogger
+
+	runtimeMux := gatewayMux()
+
+	tlsCreds, err := certs.GatewayAsClientTLSCreds(serviceConfig.GRPC.Certs)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get TLS credentials")
+	}
+	grpcEndpoint := fmt.Sprintf("dns:///%s", serviceConfig.GRPC.ListenAddress)
+
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(tlsCreds),
+		grpc.WithBlock(),
+	}
+
+	ctx, shutdownCancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer shutdownCancel()
+
+	for _, service := range services {
+		if service == readerService {
+			err = reader.RegisterReaderHandlerFromEndpoint(ctx, runtimeMux, grpcEndpoint, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if service == writerService {
+			err = writer.RegisterWriterHandlerFromEndpoint(ctx, runtimeMux, grpcEndpoint, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if service == exporterService {
+			err = exporter.RegisterExporterHandlerFromEndpoint(ctx, runtimeMux, grpcEndpoint, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
+		if service == importerService {
+			err = importer.RegisterImporterHandlerFromEndpoint(ctx, runtimeMux, grpcEndpoint, opts)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle("/", runtimeMux)
```

Comment:
Consider adding a `/openapi.json` endpoint with the spec from https://github.com/aserto-dev/openapi-directory.

See https://github.com/aserto-dev/directory/pull/292/files#diff-71c953a3cd9ce67559ba9ab35ab8237eb5c9d2d3c6767accb29d275294086471R85